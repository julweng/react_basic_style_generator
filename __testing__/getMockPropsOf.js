import { first, get, last } from "lodash"

/**
 * Gets props of a mock by its instance order or a find function.
 * @param  {component} component  class or functional component
 * @return {object}               object with selectors functions
 */
export default function getMockPropsOf(component) {
  const mock = getMockOf(component)
  if (!mock) {
    return null
  }

  const { props } = mock

  const firstInstance = () => first(props)
  const lastInstance = () => last(props)
  const nthInstance = nthIndex => get(props, getActualIndex(nthIndex))
  const findInstance = fn => props.find(fn)

  return {
    firstInstance,
    lastInstance,
    nthInstance,
    findInstance
  }
}

/**
 * Gets the mock and calls of a component.
 * @param  {object}     component  object with mock and calls references
 * @return {component}             class or functional component
 */
function getMockOf(component) {
  if (!component) {
    return null
  }

  const { mock } = component
  if (!mock) {
    return null
  }

  const { calls } = mock

  // a component is called with an array argument as [props, {}]
  // we only want the first argument which is the first element of the array
  const firstArg = args => first(args)
  const props = calls.map(firstArg)

  return {
    mock,
    calls,
    props
  }
}

/**
 * Adapts index from Jest's convention, which starts at 1.
 * See https://jestjs.io/docs/en/expect#tohavebeennthcalledwithnthcall-arg1-arg2-
 */
function getActualIndex(nthIndex) {
  return nthIndex - 1
}
