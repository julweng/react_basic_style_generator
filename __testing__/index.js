export getMockPropsOf from "./getMockPropsOf"
export mockStore from "./mockStore"
