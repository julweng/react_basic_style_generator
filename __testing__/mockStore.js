import configureMockStore from "redux-mock-store"
import thunkMiddleware from "redux-thunk"

// avoid us having to fill this in every action test file
const defaultState = {}

function mockStore(initialState = {}) {
  const state = {
    ...defaultState,
    ...initialState
  }
  return configureMockStore([thunkMiddleware])(state)
}

export default mockStore
