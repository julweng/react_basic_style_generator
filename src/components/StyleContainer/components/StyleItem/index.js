import React from "react"
import { string, shape } from "prop-types"

const StyleItem = ({ side }) => {
  return <div className={side}>Happy picking!</div>
}

StyleItem.propTypes = {
  side: string,
  style: shape({})
}

export default StyleItem
