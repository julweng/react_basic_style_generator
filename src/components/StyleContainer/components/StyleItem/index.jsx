import React from "react"
import { string, shape, arrayOf, func } from "prop-types"
import { Selection } from "../index"
import "./style.less"

const StyleItem = ({ side, style, fonts, font, updateFont }) => {
  const defaultText = "Happy Picking"

  return (
    <div className={side} style={style}>
      <div className="selection__container">
        <Selection
          type="font"
          side={side}
          fonts={fonts}
          font={font}
          updateFont={updateFont}
        />
        <Selection type="palette" />
      </div>
      <p>{font.family || defaultText}</p>
    </div>
  )
}

StyleItem.propTypes = {
  side: string,
  style: shape({}),
  color: shape({}),
  font: shape({}),
  fonts: arrayOf(shape({})),
  updateFont: func,
  updateColor: func
}

export default StyleItem
