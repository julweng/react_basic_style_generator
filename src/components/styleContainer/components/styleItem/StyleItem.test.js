import React from "react"
import { render, cleanup } from "@testing-library/react"
import { getMockPropsOf } from "__testing__"
import { Selection, StyleRule } from "../index"
import "jest-dom/extend-expect"
import StyleItem from "./index"

jest.mock("../Selection", () => jest.fn(() => <div>SelectionMock</div>))

jest.mock("../StyleRule", () => jest.fn(() => <div>StyleRuleMock</div>))

const createProps = ({
  style = {
    color: "#000",
    backgroundColor: "#FFF",
    fontFamily: "Arial, sans-serif",
    fontWeight: 400,
    fontStyle: "normal"
  },
  fonts = [
    {
      family: "Arial",
      category: "sans-serif",
      variants: ["regular", "100italic"]
    }
  ],
  font = {
    family: "Arial",
    category: "sans-serif",
    variant: { regular: "regular" }
  },
  side = "left"
} = {}) => ({
  style,
  side,
  fonts,
  font,
  updateFont: jest.fn(),
  updateColor: jest.fn()
})

describe("StyleItem", () => {
  afterEach(cleanup)

  it("Should render without crashing", () => {
    const props = createProps()
    const { container } = render(<StyleItem {...props} />)
    expect(container).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("Should render first child container with class name based on the side prop", () => {
    const props = createProps()
    const { container } = render(<StyleItem {...props} />)
    expect(container.firstChild).toHaveClass("left")
  })

  it("Should render first child with style prop", () => {
    const props = createProps()
    const { container } = render(<StyleItem {...props} />)
    expect(container.firstChild).toHaveStyle(`
      color: "#000";
      backgroundColor: "#FFF";
      fontFamily: "Arial, sans-serif";
      fontWeight: 400;
      fontStyle: "normal";
    `)
  })

  it("Should render two instances of Selection", () => {
    const props = createProps()
    const { getAllByText } = render(<StyleItem {...props} />)
    const mockSelections = getAllByText("SelectionMock")
    expect(mockSelections.length).toEqual(2)
  })

  it("Should render a selection instance of the type font", () => {
    const props = createProps()
    render(<StyleItem {...props} />)
    const passedProps = getMockPropsOf(Selection).firstInstance()

    const { side, fonts, font, updateFont } = props

    expect(passedProps).toEqual({
      type: "font",
      side,
      fonts,
      font,
      updateFont
    })
  })

  it("Should render a selection instance of the type palette", () => {
    const props = createProps()
    render(<StyleItem {...props} />)
    const passedProps = getMockPropsOf(Selection).lastInstance()

    const { side, style, updateColor } = props

    expect(passedProps).toEqual({
      type: "palette",
      side,
      color: style,
      updateColor
    })
  })

  it("Should display a default message when font is not available", () => {
    const props = createProps({ font: {} })
    const { container } = render(<StyleItem {...props} />)
    const msg = container.querySelector("p")
    expect(msg).toHaveTextContent("Happy Picking")
  })

  it("Should display font family when font is available", () => {
    const props = createProps()
    const { container } = render(<StyleItem {...props} />)
    const msg = container.querySelector("p")
    expect(msg).toHaveTextContent("Arial")
  })

  it("Should render an instance of StyleRule", () => {
    const props = createProps()
    const { getAllByText } = render(<StyleItem {...props} />)
    const mockStyleRule = getAllByText("StyleRuleMock")
    expect(mockStyleRule.length).toEqual(1)
  })

  it("Should pass the style prop to StyleRule", () => {
    const props = createProps()
    render(<StyleItem {...props} />)
    const passedProps = getMockPropsOf(StyleRule).firstInstance()
    expect(passedProps).toEqual({ style: props.style })
  })
})
