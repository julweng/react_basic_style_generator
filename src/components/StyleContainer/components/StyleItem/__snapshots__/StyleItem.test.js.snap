// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleItem Should render without crashing 1`] = `
<div>
  <div
    class="left"
    style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255); font-family: Arial, sans-serif; font-weight: 400; font-style: normal;"
  >
    <div
      class="selection__container"
    >
      <div>
        SelectionMock
      </div>
      <div>
        SelectionMock
      </div>
    </div>
    <p>
      Arial
    </p>
    <div>
      StyleRuleMock
    </div>
  </div>
</div>
`;
